apps/heft/src/pluginFramework/logging/ScopedLogger.ts

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { Terminal, ITerminalProvider } from '@rushstack/node-core-library';
+
+import { IHeftPlugin } from '../IHeftPlugin';
+import { PrefixProxyTerminalProvider } from '../../utilities/PrefixProxyTerminalProvider';
+
+export interface IScopedLoggerOptions {
+  requestingPlugin: IHeftPlugin;
+  loggerName: string;
+  terminalProvider: ITerminalProvider;
+  getShouldPrintStacks: () => boolean;
+}
+
+/**
+ * @public
+ */
+export interface IScopedLogger {
+  readonly terminal: Terminal;
+
+  /**
+   * Call this function to emit an error to the heft runtime.
+   */
+  emitError(error: Error): void;
+
+  /**
+   * Call this function to emit an warning to the heft runtime.
+   */
+  emitWarning(warning: Error): void;
+}
+
+/**
+ * @public
+ */
+export class ScopedLogger implements IScopedLogger {
+  private readonly _options: IScopedLoggerOptions;
+  private readonly _errors: Error[] = [];
+  private readonly _warnings: Error[] = [];
+
+  private get _shouldPrintStacks(): boolean {
+    return this._options.getShouldPrintStacks();
+  }
+
+  public get errors(): ReadonlyArray<Error> {
+    return [...this._errors];
+  }
+
+  public get warnings(): ReadonlyArray<Error> {
+    return [...this._warnings];
+  }
+
+  /**
+   * @internal
+   */
+  public readonly _requestingPlugin: IHeftPlugin;
+
+  public readonly loggerName: string;
+
+  public readonly terminalProvider: ITerminalProvider;
+
+  public readonly terminal: Terminal;
+
+  /**
+   * @internal
+   */
+  public constructor(options: IScopedLoggerOptions) {
+    this._options = options;
+    this._requestingPlugin = options.requestingPlugin;
+    this.loggerName = options.loggerName;
+
+    this.terminalProvider = new PrefixProxyTerminalProvider(
+      options.terminalProvider,
+      `[${this.loggerName}] `
+    );
+    this.terminal = new Terminal(this.terminalProvider);
+  }
+
+  /**
+   * {@inheritdoc IScopedLogger.emitError}
+   */
+  public emitError(error: Error): void {
+    this._errors.push(error);
+    this.terminal.writeErrorLine(`ERROR: ${error.message}`);
```

Comment:
I pushed a small change that fixes some lint "errors" to be warnings. More to come in a separate PR.

Repro:
1. Go to `build-tests\heft-node-basic-test\src\index.ts`
2. Add `let x;` to the bottom of that file.
3. Run `rushx build`

You will see this output:
```
[typescript] Encountered 1 TypeScript error:
[typescript] ERROR: src\index.ts:9:5 - (TS6133) 'x' is declared but its value is never read.
[eslint] Encountered 2 ESLint errors:
[eslint] WARNING: src\index.ts:9:5 - (@typescript-eslint/typedef) Expected x to have a type annotation.
[eslint] WARNING: src\index.ts:9:5 - (@typescript-eslint/no-unused-vars) 'x' is defined but never used.
 ---- Compile finished (2828ms) ----
-------------------- Finished (3.781s) --------------------
Encountered 2 warnings:
  [eslint] src\index.ts:9:5 - (@typescript-eslint/typedef) Expected x to have a type annotation.
  [eslint] src\index.ts:9:5 - (@typescript-eslint/no-unused-vars) 'x' is defined but never used.
Encountered 1 errors:
  [typescript] src\index.ts:9:5 - (TS6133) 'x' is declared but its value is never read.
```

Feedback:
- `WARNING` should not be yelling in all caps.  Make it `Warning`
- I'm less sure about `ERROR` but it also seems like yelling. I'm thinking the ideal should be to ensure that the word `Error` appears in logs ONLY when an actual error occurs, thus it doesn't need all caps. I myself have wasted time investigating an "error" log message that is actually expected output in a successful build.  These waste could have been avoided if the toolchain forced a failure whenever that string appears in a log message (e.g. the Jest test that does `console.log("Error: I validated that logging works, yay the test passed!");`.  Other people might have other opinions though.
- `Encountered 2 ESLint errors` should say "warnings" or maybe "issues" if we can't know.
- Why is the compiler classifying `TS6133`  as a hard error? We should investigate that.

Reply:
I want to normalize the way path errors/warnings are handled in a separate PR.